File: 0501-1000/0670. Maximum Swap.java
// Medium
// Greedy
// Time:O(n),Space:O(1)
// https://leetcode.cn/problems/maximum-swap/

class Solution {
    public int maximumSwap(int num) {
        String nums = Integer.toString(num);
        char[] digit = nums.toCharArray();
        int n = digit.length;

        // store last index of each digit(0-9)
        int[] last = new int[10];
        for (int i = 0; i < n; i++) {
            last[digit[i] - '0'] = i;
        }

        for (int i = 0; i < n; i++) {
            // check if a larger digit exists to the right of cur digit
            for (int d = 9; d > digit[i] - '0'; d--) {
                if (last[d] > i) {
                    char tmp = digit[i];
                    digit[i] = digit[last[d]];
                    digit[last[d]] = tmp;
                    return Integer.parseInt(new String(digit));
                }
            }
        }
        return num; // if no swap
    }
}

File: 0501-1000/0670. Maximum Swap.py
# Time:O(n), Space:O(n)

class Solution:
    def maximumSwap(self, num: int) -> int:
        s = list(str(num))
        n = len(s)

        last = {int(d): i for i, d in enumerate(s)}

        for i in range(n):
            digit = int(s[i])
            for d in range(9, digit, -1):
                if d in last and last[d] > i:
                    s[i], s[last[d]] = s[last[d]], s[i]
                    return int("".join(s))

        return num
